lib/push.py

```diff
@@ -149,6 +191,33 @@ def prepare(self):
          if (i != non_empty_winner):
             hist[i]["empty_layer"] = True
       config["history"] = hist
+
+      (sid, git_hash) = bu.cache.find_image(self.image)
+      ul_ref = str(ch.storage.upload_cache)
+      conf_path = ul_ref + "/" + git_hash + ".config.json"
+      man_path = ul_ref + "/" + git_hash + ".manifest.json"
+      with open(Path(conf_path), "w") as conf:
+         json.dump(config, conf)
+      with open(Path(man_path), "w") as man:
+         json.dump(manifest, man)
+      self.layers = tars_c
+      return {"config": config, "manifest": manifest}
+
+   def prepare(self):
+      """Prepare self.image for pushing to self.dst_ref. Return tuple: (list
+         of gzipped layer tarball paths, config as a sequence of bytes,
+         manifest as a sequence of bytes).
+
+         There is not currently any support for re-using any previously
+         prepared files already in the upload cache, because we don't yet have
+         a way to know if these have changed until they are already build."""
+      # TODO: Check if build cache is even enabled first before below
+      # CHECK: use prev_prepeared to check if git-head.conf, etc. exists
+      prev_prepared = self.get_prev_prepared()
+      if prev_prepared is None:
+         prev_prepared = self.prepare_new()
```

Comment:
I'm not sure this logic needs to be split into three methods. Consider putting all the logic here in `prepare()`.

Reply:
I put some more thought into the structure of this, and actually ended up going further in the opposite way. I agree it was a bit oddly structured how I had it, but rather than put all logic in ```prepare()``` I have split it up a bit further. It is now structured similar to ```push()```; meaning each of aspects of preparing to push (as I view it) are now independent functions. These aspects are:

1. check for existing config, manifest, and tar
    a. If not present, prepare new versions of these files
2. Store config bytes, manifest bytes, and layers

If this still seems worse than putting all logic in ```push()``` I am happy to change it back. This is the way that seemed slightly best to me, but I am curious what you think.